### PR TITLE
[MM-47181] Migrate "components/suggestion/at_mention_provider" and associated tests to Typescript

### PR DESCRIPTION
**Pull request by Sudhanva-Nadiger**
#### Summary

- Migrate "components/suggestion/at_mention_provider" and associated tests to Typescript 
- resolves: #23406

#### Ticket Link
- JIRA: https://mattermost.atlassian.net/browse/MM-47181

#### Screenshots
image

#### Release Note

```release-note
None
```

### CODE REVIEW

**Comment by M-ZubairAhmed** [comment]
@Sudhanva-Nadiger can you please check the failing CI

**Comment by Sudhanva-Nadiger** [comment]
Yes ! will do that!

**Comment by Sudhanva-Nadiger** [comment]
@M-ZubairAhmed all checks have passed ! its ready to review :)

File: webapp/channels/src/components/suggestion/at_mention_provider/at_mention_provider.tsx [hunk 1]
```diff
@@ -342,15 +398,17 @@ export default class AtMentionProvider extends Provider {
...
             }
 
-            this.updateMatches(resultCallback, this.items().concat([{
+            this.updateMatches(resultCallback, [...this.items(), ...[{
                 type: Constants.MENTION_MORE_MEMBERS,
                 loading: true,
-            }]));
+            }]]);
+
+            this.items().concat();
```
Comment (M-ZubairAhmed) [inline webapp/channels/src/components/suggestion/at_mention_provider/at_mention_provider.tsx:411]: why do we need this?

File: webapp/channels/src/components/suggestion/at_mention_provider/at_mention_provider.tsx
Comment (M-ZubairAhmed) [inline webapp/channels/src/components/suggestion/at_mention_provider/at_mention_provider.tsx:1]: I am concerned with the concat removal from here. Was it causing a type issue which is not fixable?

**Review by M-ZubairAhmed — CHANGES_REQUESTED** [review]

File: webapp/channels/src/components/suggestion/at_mention_provider/at_mention_provider.tsx (hunk 1 above)
Comment (Sudhanva-Nadiger) [reply webapp/channels/src/components/suggestion/at_mention_provider/at_mention_provider.tsx:411]: i might have forgot to remove that ! will remove that !

File: webapp/channels/src/components/suggestion/at_mention_provider/at_mention_provider.tsx
Comment (M-ZubairAhmed) [reply webapp/channels/src/components/suggestion/at_mention_provider/at_mention_provider.tsx:1]: @Sudhanva-Nadiger let me know when you have pushed the changes for this we can review again

File: webapp/channels/src/components/suggestion/at_mention_provider/at_mention_provider.test.tsx [hunk 2]
```diff
@@ -359,7 +361,7 @@ describe('components/suggestion/at_mention_provider/AtMentionProvider', () => {
         };
 
         const provider = new AtMentionProvider(params);
-        jest.spyOn(provider, 'getProfilesWithLastViewAtInChannel').mockImplementation(() => [userid10, userid3, userid1, userid2, userid12]);
+        jest.spyOn(provider, 'getProfilesWithLastViewAtInChannel').mockImplementation(() => [userid10, userid3, userid1, userid2]);
```
Comment (jespino) [inline webapp/channels/src/components/suggestion/at_mention_provider/at_mention_provider.test.tsx:364]: Why are removing here the `userid12` from this list?

File: webapp/channels/src/components/suggestion/at_mention_provider/at_mention_provider.tsx [hunk 3]
```diff
@@ -294,11 +350,11 @@ export default class AtMentionProvider extends Provider {
             filter((member) => !localUserIds[member.id]).
             sort(orderUsers);
 
-        return priorityProfiles.concat(localAndRemoteMembers).concat(localAndRemoteGroups).concat(specialMentions).concat(remoteNonMembers);
+        return [...priorityProfiles, ...localAndRemoteMembers, ...localAndRemoteGroups, ...specialMentions, ...remoteNonMembers];
```
Comment (M-ZubairAhmed) [inline webapp/channels/src/components/suggestion/at_mention_provider/at_mention_provider.tsx:353]: @Sudhanva-Nadiger were you able to revert these changes?

**Review by M-ZubairAhmed — CHANGES_REQUESTED** [review]

File: webapp/channels/src/components/suggestion/at_mention_provider/at_mention_provider.tsx (hunk 3 above)
Comment (Sudhanva-Nadiger) [reply webapp/channels/src/components/suggestion/at_mention_provider/at_mention_provider.tsx:353]: Concat gives error ! Because there is multiple types of array are getting concatenated

File: webapp/channels/src/components/suggestion/at_mention_provider/at_mention_provider.test.tsx (hunk 2 above)
Comment (Sudhanva-Nadiger) [reply webapp/channels/src/components/suggestion/at_mention_provider/at_mention_provider.test.tsx:364]: Oh my bad ! I will add that back ! It was not intensional

File: webapp/channels/src/components/suggestion/at_mention_provider/at_mention_provider.tsx (hunk 3 above)
Comment (M-ZubairAhmed) [reply webapp/channels/src/components/suggestion/at_mention_provider/at_mention_provider.tsx:353]: i see, i think its fine then

**Comment by M-ZubairAhmed** [comment]
adding @hmhealey since he has a better understanding of this component

**Comment by M-ZubairAhmed** [comment]
@hmhealey can you please take a look at this PR when you get time

**Comment by hmhealey** [comment]
Yep, I'll take a look. Sorry that I managed to forget about this one since I remember seeing it at some point

File: webapp/channels/src/components/textbox.test.tsx [hunk 4]
```diff
@@ -28,106 +26,118 @@ describe('components/TextBox', () => {
...
+            ...baseProps,
+            id: 'someid',
+            value: 'some test text',
+            onChange: emptyFunction,
+            onKeyPress: emptyFunction,
+            characterLimit: 4000,
+            createMessage: 'placeholder text',
+            supportsCommands: false,
+            rootId: 'root_id',
+            onComposition: jest.fn().mockReturnValue({}),
```
Comment (hmhealey) [inline webapp/channels/src/components/textbox.test.tsx:68]: This is technically changing the behaviour of these tests slightly since these functions now return an empty object instead of nothing, but since the tests aren't breaking, that's not a problem

**Review by hmhealey — APPROVED** [review]
Thanks for the PR, @Sudhanva-Nadiger. This looks good to me! Good job dealing with this code since it's definitely not easy to type, particularly because we mix so many weird slighty-different types in this provider

**Comment by hmhealey** [comment]
Updating this to the latest branch since I've overhauled a bunch of type definitions recently. Testing locally, everything seems to pass still though

/update-branch

File: webapp/channels/src/components/textbox.test.tsx (hunk 4 above)
Comment (M-ZubairAhmed) [reply webapp/channels/src/components/textbox.test.tsx:68]: we will run e2e to be sure we didnt break any of them

**Review by yasserfaraazkhan — APPROVED** [review]

**Review by M-ZubairAhmed — APPROVED** [review]
